Upstream: http://lists.gnu.org/archive/html/grub-devel/2009-09/msg00056.html
Description: Don't display introductory message about line editing unless
 we're actually offering a shell prompt.
UbuntuSpecific: This is believed to be a workaround for a different bug.
 We'll go with this for now, but will drop this in favour of a better fix
 upstream if somebody figures out what that is.

diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta1/normal/main.c grub2-1.97~beta1.new/normal/main.c
--- grub2-1.97~beta1/normal/main.c	2009-08-30 17:02:11.000000000 +0100
+++ grub2-1.97~beta1.new/normal/main.c	2009-09-03 13:52:43.000000000 +0100
@@ -503,7 +503,7 @@
 }
 
 static grub_err_t
-grub_normal_reader_init (void)
+grub_normal_display_intro (void)
 {
   grub_normal_init_page ();
   grub_setcursor (1);
@@ -524,6 +524,13 @@
 {
   grub_parser_t parser = grub_parser_get_current ();
   char prompt[8 + grub_strlen (parser->name)];
+  static int displayed_intro;
+
+  if (! displayed_intro)
+    {
+      grub_normal_display_intro ();
+      displayed_intro = 1;
+    }
 
   grub_sprintf (prompt, "%s:%s> ", parser->name, (cont) ? "" : "grub");
 
@@ -547,7 +554,6 @@
 static struct grub_reader grub_normal_reader =
   {
     .name = "normal",
-    .init = grub_normal_reader_init,
     .read_line = grub_normal_read_line
   };
 
